arbor/partition_load_balance.cpp

```diff
@@ -1,234 +1,219 @@
+}
 
-
-    // Create a flat vector of the cell kinds present on this node,
-    // partitioned such that kinds for which GPU implementation are
-    // listed before the others. This is a very primitive attempt at
-    // scheduling; the cell_groups that run on the GPU will be executed
-    // before other cell_groups, which is likely to be more efficient.
-    //
-    // TODO: This creates an dependency between the load balancer and
-    // the threading internals. We need support for setting the priority
-    // of cell group updates according to rules such as the back end on
-    // which the cell group is running.
-
-    auto has_gpu_backend = [&ctx](cell_kind c) {
-        return cell_kind_supported(c, backend_kind::gpu, *ctx);
-    };
-
-    std::vector<cell_kind> kinds;
-    for (auto l: kind_lists) {
-        kinds.push_back(cell_kind(l.first));
+struct group_parameters {
+    cell_kind kind;
+    backend_kind backend;
+    size_t size;
+};
+
+// Create a flat vector of the cell kinds present on this node, sorted such that
+// kinds for which GPU implementation are listed before the others. This is a
+// very primitive attempt at scheduling; the cell_groups that run on the GPU
+// will be executed before other cell_groups, which is likely to be more
+// efficient.
+//
+// TODO: This creates an dependency between the load balancer and the threading
+// internals. We need support for setting the priority of cell group updates
+// according to rules such as the back end on which the cell group is running.
+auto build_group_parameters(context ctx,
+                            const partition_hint_map& hint_map,
+                            const std::unordered_map<cell_kind, std::vector<cell_gid_type>>& kind_lists) {
+    std::vector<group_parameters> res;
+    for (const auto& [kind, _gids]: kind_lists) {
+        const auto& [backend, group_size] = get_backend(ctx, kind, hint_map);
+        res.push_back({kind, backend, group_size});
     }
-    std::partition(kinds.begin(), kinds.end(), has_gpu_backend);
+    util::sort_by(res, [](const auto& p) { return p.kind; });
+    return res;
+}
 
-    std::vector<group_description> groups;
-    for (auto k: kinds) {
-        partition_hint hint;
-        if (auto opt_hint = util::value_by_key(hint_map, k)) {
-            hint = opt_hint.value();
-            if(!hint.cpu_group_size) {
-                throw arbor_exception(arb::util::pprintf("unable to perform load balancing because {} has invalid suggested cpu_cell_group size of {}", k, hint.cpu_group_size));
-            }
-            if(hint.prefer_gpu && !hint.gpu_group_size) {
-                throw arbor_exception(arb::util::pprintf("unable to perform load balancing because {} has invalid suggested gpu_cell_group size of {}", k, hint.gpu_group_size));
-            }
-        }
+// Build the list of GJ-connected cells local to this domain.
+// NOTE We put this into its own function to avoid increasing RSS.
```

Comment:
can you elaborate on why this reduces RSS?

Reply:
By putting temporary data structures into their own scope we ensure that they are de-allocated
before the next set of temporaries comes into being.